sql/src/main/java/io/crate/execution/dml/upsert/InsertSourceFromCells.java

```diff
@@ -44,74 +47,118 @@
 
 import javax.annotation.Nullable;
 import java.io.IOException;
-import java.util.HashMap;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
 import java.util.List;
+import java.util.Map;
 
 public final class InsertSourceFromCells implements InsertSourceGen {
 
     private final List<Reference> targets;
     private final ArrayRow row = new ArrayRow();
     private final CheckConstraints<Row, CollectExpression<Row, ?>> checks;
     private final GeneratedColumns<Row> generatedColumns;
+    private final List<Object> defaultValues;
 
     InsertSourceFromCells(TransactionContext txnCtx,
                           Functions functions,
                           DocTableInfo table,
                           String indexName,
                           GeneratedColumns.Validation validation,
                           List<Reference> targets) {
-        this.targets = targets;
-        ReferencesFromInputRow referenceResolver = new ReferencesFromInputRow(targets, table.partitionedByColumns(), indexName);
+        // Ensure targets are mutable because defaults may be added
+        this.targets = new ArrayList<>(targets);
```

Comment:
Is this safe for re-use?

In the `TransportShardUpsertAction` we create a single `InsertSourceFromCells` instance and then use it for multiple params. 

What if there is a null value early and later on there isn't?

Reply:
Should be safe yes, its only mutated once to add missing columns with a default constraint.  The default value of a target table column with a default constraint will be applied if the column is not explicitly defined for writing. Everything will be the same for all rows.

Example:
```
CREATE TABLE t1 (id int, x int);
CREATE TABLE t2 (id int, x int default 1, y as x + 1);
INSERT INTO t1 (id, x) VALUES (1, 3);
INSERT INTO t2 (id) (SELECT id, x FROM t1);         -> t2.x = 1 (default) for all rows
INSERT INTO t2 (id, x) (SELECT id, x FROM t1);      -> t2.x = t1.x for all rows
```

Reply:
I've pushed a fixup which uses a slightly different pattern and also avoids array copies by using a new `BiArrayRow`.